#GRUPO 3

OSCAR HERNAN	ESTRADA VARGAS

CAMILO	ESTUPIÑAN OROZCO

YASMIN	GARCIA GAVIRIA

CARLOS ENRIQUE	JARAMILLO AROS


In [275]:
# Importamos las librerías 

import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
from matplotlib import rcParams
import seaborn as sns

# Estilos 

rcParams['figure.figsize'] = (12,  6)

In [276]:
# Cargamos el dataset de prueba
# URL del dataset
filename = "https://raw.githubusercontent.com/armandoordonez/eda_couse/main/data/Car_details_v3.csv"

# Leer el archivo CSV
df = pd.read_csv(filename)

In [277]:
df.head()

name  year  selling_price  km_driven    fuel  \
0        Maruti Swift Dzire VDI  2014         450000     145500  Diesel   
1  Skoda Rapid 1.5 TDI Ambition  2014         370000     120000  Diesel   
2      Honda City 2017-2020 EXi  2006         158000     140000  Petrol   
3     Hyundai i20 Sportz Diesel  2010         225000     127000  Diesel   
4        Maruti Swift VXI BSIII  2007         130000     120000  Petrol   

  seller_type transmission         owner     mileage   engine   max_power  \
0  Individual       Manual   First Owner   23.4 kmpl  1248 CC      74 bhp   
1  Individual       Manual  Second Owner  21.14 kmpl  1498 CC  103.52 bhp   
2  Individual       Manual   Third Owner   17.7 kmpl  1497 CC      78 bhp   
3  Individual       Manual   First Owner   23.0 kmpl  1396 CC      90 bhp   
4  Individual       Manual   First Owner   16.1 kmpl  1298 CC    88.2 bhp   

                     torque  seats  
0            190Nm@ 2000rpm    5.0  
1       250Nm@ 1500-2500rpm    5.0  
2     12.7@ 2,700(kgm@ rpm)    5.0  
3  22.4 kgm at 1750-2750rpm    5.0  
4     11.5@ 4,500(kgm@ rpm)    5.0

In [278]:
df.tail()

name  year  selling_price  km_driven    fuel  \
8123       Hyundai i20 Magna  2013         320000     110000  Petrol   
8124   Hyundai Verna CRDi SX  2007         135000     119000  Diesel   
8125  Maruti Swift Dzire ZDi  2009         382000     120000  Diesel   
8126         Tata Indigo CR4  2013         290000      25000  Diesel   
8127         Tata Indigo CR4  2013         290000      25000  Diesel   

     seller_type transmission                 owner     mileage   engine  \
8123  Individual       Manual           First Owner   18.5 kmpl  1197 CC   
8124  Individual       Manual  Fourth & Above Owner   16.8 kmpl  1493 CC   
8125  Individual       Manual           First Owner   19.3 kmpl  1248 CC   
8126  Individual       Manual           First Owner  23.57 kmpl  1396 CC   
8127  Individual       Manual           First Owner  23.57 kmpl  1396 CC   

      max_power                     torque  seats  
8123  82.85 bhp           113.7Nm@ 4000rpm    5.0  
8124    110 bhp  24@ 1,900-2,750(kgm@ rpm)    5.0  
8125   73.9 bhp             190Nm@ 2000rpm    5.0  
8126     70 bhp        140Nm@ 1800-3000rpm    5.0  
8127     70 bhp        140Nm@ 1800-3000rpm    5.0

In [279]:
df.shape

(8128, 13)

In [280]:
df.info()
#Revisar duplicados
#Revisar tipos

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8128 entries, 0 to 8127
Data columns (total 13 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   name           8128 non-null   object 
 1   year           8128 non-null   int64  
 2   selling_price  8128 non-null   int64  
 3   km_driven      8128 non-null   int64  
 4   fuel           8128 non-null   object 
 5   seller_type    8128 non-null   object 
 6   transmission   8128 non-null   object 
 7   owner          8128 non-null   object 
 8   mileage        7907 non-null   object 
 9   engine         7907 non-null   object 
 10  max_power      7913 non-null   object 
 11  torque         7906 non-null   object 
 12  seats          7907 non-null   float64
dtypes: float64(1), int64(3), object(9)
memory usage: 825.6+ KB


In [281]:
df.duplicated().sum()

1202

In [282]:
#Revisar perdidos

In [283]:
Value_count = df.isna().sum()
Value_Count_Mayor = Value_count[Value_count > 0]
print(f'Resumen:\r\n',Value_Count_Mayor)

Resumen:
 mileage      221
engine       221
max_power    215
torque       222
seats        221
dtype: int64


In [284]:
# Promedio/media de la columna seats para reemplazarlos.
avg_seats = df["seats"].astype("float").mean().round()
print("Promedio de seats:", avg_seats)
df["seats"].replace(np.nan, avg_seats, inplace=True)

Promedio de seats: 5.0


In [285]:
Value_count = df.isna().sum()
Value_Count_Mayor = Value_count[Value_count > 0]
print(f'Resumen:\r\n',Value_Count_Mayor)

Resumen:
 mileage      221
engine       221
max_power    215
torque       222
dtype: int64


In [286]:
df.dropna(subset=["mileage"], axis=0, inplace=True)
df.reset_index(drop=True, inplace=True)

In [287]:
df.dropna(subset=["engine"], axis=0, inplace=True)
df.reset_index(drop=True, inplace=True)

In [288]:
Value_count = df.isna().sum()
Value_Count_Mayor = Value_count[Value_count > 0]
print(f'Resumen:\r\n',Value_Count_Mayor)

Resumen:
 torque    1
dtype: int64


In [289]:
df.dropna(subset=["torque"], axis=0, inplace=True)
df.reset_index(drop=True, inplace=True)

In [290]:
Value_count = df.isna().sum()
Value_Count_Mayor = Value_count[Value_count > 0]
print(f'Resumen:\r\n',Value_Count_Mayor)

Resumen:
 Series([], dtype: int64)


In [291]:
#Bajaron los duplicados?
df.duplicated().sum()


1189

In [292]:
#Bajo la cantidad de registros ?
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7906 entries, 0 to 7905
Data columns (total 13 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   name           7906 non-null   object 
 1   year           7906 non-null   int64  
 2   selling_price  7906 non-null   int64  
 3   km_driven      7906 non-null   int64  
 4   fuel           7906 non-null   object 
 5   seller_type    7906 non-null   object 
 6   transmission   7906 non-null   object 
 7   owner          7906 non-null   object 
 8   mileage        7906 non-null   object 
 9   engine         7906 non-null   object 
 10  max_power      7906 non-null   object 
 11  torque         7906 non-null   object 
 12  seats          7906 non-null   float64
dtypes: float64(1), int64(3), object(9)
memory usage: 803.1+ KB


### Se decide no borrar los registros por criterios.

Por un criterio conjunto se decide no borrar los registros duplicados, pues no tenemos la definicion completa de como se tomaron los datos y es posible que las columnas no estén normalizadas, por tal motivo estos duplicados sea porque faltan columnas como la placa del vehiculo por ejemplo, y teniendo en cuenta que al final se quiere revisar precios de ventas, estos duplicados tengan mucho que ver en calculos totates si se deciden borrar los 1202 registros.

```df.drop_duplicates(keep='first')``` <- NO

In [293]:
df['engine'].unique()

array(['1248 CC', '1498 CC', '1497 CC', '1396 CC', '1298 CC', '1197 CC',
       '1061 CC', '796 CC', '1364 CC', '1399 CC', '1461 CC', '993 CC',
       '1198 CC', '1199 CC', '998 CC', '1591 CC', '2179 CC', '1368 CC',
       '2982 CC', '2494 CC', '2143 CC', '2477 CC', '1462 CC', '2755 CC',
       '1968 CC', '1798 CC', '1196 CC', '1373 CC', '1598 CC', '1998 CC',
       '1086 CC', '1194 CC', '1172 CC', '1405 CC', '1582 CC', '999 CC',
       '2487 CC', '1999 CC', '3604 CC', '2987 CC', '1995 CC', '1451 CC',
       '1969 CC', '2967 CC', '2497 CC', '1797 CC', '1991 CC', '2362 CC',
       '1493 CC', '1599 CC', '1341 CC', '1794 CC', '799 CC', '1193 CC',
       '2696 CC', '1495 CC', '1186 CC', '1047 CC', '2498 CC', '2956 CC',
       '2523 CC', '1120 CC', '624 CC', '1496 CC', '1984 CC', '2354 CC',
       '814 CC', '793 CC', '1799 CC', '936 CC', '1956 CC', '1997 CC',
       '1499 CC', '1948 CC', '2997 CC', '2489 CC', '2499 CC', '2609 CC',
       '2953 CC', '1150 CC', '1994 CC', '1388 CC', '1527 CC'

### Preparacion de datos

In [294]:
#### Vamos a analizar como limpar los datos
#milage revisaremos las caracteristicas de las unidades de medida que tiene cada uno
#https://datascience.stackexchange.com/questions/55663/one-feature-several-units

#CNG: ['km/kg']
#Diesel: ['kmpl']
#Petrol: ['kmpl']
#LPG: ['km/kg']
#Electric: [nan]

In [295]:
unique_values, counts = np.unique(df['mileage'].values, return_counts=True)
unique_df = pd.DataFrame({'Value': unique_values, 'Count': counts})
print(unique_df)

           Value  Count
0       0.0 kmpl     17
1      10.0 kmpl      2
2      10.1 kmpl      2
3      10.5 kmpl     17
4     10.71 kmpl      1
..           ...    ...
388   33.0 km/kg      1
389  33.44 km/kg      3
390    42.0 kmpl      1
391     9.0 kmpl      4
392     9.5 kmpl      6

[393 rows x 2 columns]


In [296]:
#Contemos cada uno
df[df['mileage'].str.contains('km/kg', case=False)]

name  year  selling_price  km_driven  \
6               Maruti Wagon R LXI DUO BSIII  2007          96000     175000   
33          Maruti Alto 800 CNG LXI Optional  2019         330000      10000   
86               Maruti Wagon R LXI LPG BSIV  2010         225000      44000   
196                   Maruti Wagon R LXI CNG  2016         360000      50000   
388              Maruti Eeco CNG 5 Seater AC  2020         409999      35000   
...                                      ...   ...            ...        ...   
7107             Maruti Alto Green LXi (CNG)  2010         155000      70000   
7332        Maruti Alto 800 CNG LXI Optional  2019         330000      10000   
7428              Chevrolet Spark 1.0 LT LPG  2012         300000      70000   
7517        Maruti Eeco CNG 5 Seater AC BSIV  2015         380000      20000   
7695  Maruti Eeco 5 STR With AC Plus HTR CNG  2011         209000          1   

     fuel seller_type transmission                 owner      mileage  \
6     LPG  Individual       Manual           First Owner   17.3 km/kg   
33    CNG  Individual       Manual          Second Owner  33.44 km/kg   
86    LPG      Dealer       Manual           First Owner   26.2 km/kg   
196   CNG  Individual       Manual           First Owner   26.6 km/kg   
388   CNG  Individual       Manual           First Owner  20.88 km/kg   
...   ...         ...          ...                   ...          ...   
7107  CNG  Individual       Manual          Second Owner  26.83 km/kg   
7332  CNG  Individual       Manual          Second Owner  33.44 km/kg   
7428  LPG  Individual       Manual           First Owner   13.2 km/kg   
7517  CNG  Individual       Manual           First Owner  21.94 km/kg   
7695  CNG  Individual       Manual  Fourth & Above Owner   15.1 km/kg   

       engine  max_power                torque  seats  
6     1061 CC   57.5 bhp  7.8@ 4,500(kgm@ rpm)    5.0  
33     796 CC   40.3 bhp         60Nm@ 3500rpm    4.0  
86     998 CC   58.2 bhp         77Nm@ 3500rpm    5.0  
196    998 CC  58.16 bhp         77Nm@ 3500rpm    5.0  
388   1196 CC   61.7 bhp         85Nm@ 3000rpm    5.0  
...       ...        ...                   ...    ...  
7107   796 CC   38.4 bhp         62Nm@ 3000rpm    5.0  
7332   796 CC   40.3 bhp         60Nm@ 3500rpm    4.0  
7428   995 CC   60.2 bhp       90.3Nm@ 4200rpm    5.0  
7517  1196 CC   61.7 bhp         85Nm@ 3000rpm    5.0  
7695  1196 CC     73 bhp        101Nm@ 3000rpm    5.0  

[87 rows x 13 columns]

In [297]:
df[df['mileage'].str.contains('kmpl', case=False)]

name  year  selling_price  km_driven    fuel  \
0           Maruti Swift Dzire VDI  2014         450000     145500  Diesel   
1     Skoda Rapid 1.5 TDI Ambition  2014         370000     120000  Diesel   
2         Honda City 2017-2020 EXi  2006         158000     140000  Petrol   
3        Hyundai i20 Sportz Diesel  2010         225000     127000  Diesel   
4           Maruti Swift VXI BSIII  2007         130000     120000  Petrol   
...                            ...   ...            ...        ...     ...   
7901             Hyundai i20 Magna  2013         320000     110000  Petrol   
7902         Hyundai Verna CRDi SX  2007         135000     119000  Diesel   
7903        Maruti Swift Dzire ZDi  2009         382000     120000  Diesel   
7904               Tata Indigo CR4  2013         290000      25000  Diesel   
7905               Tata Indigo CR4  2013         290000      25000  Diesel   

     seller_type transmission                 owner     mileage   engine  \
0     Individual       Manual           First Owner   23.4 kmpl  1248 CC   
1     Individual       Manual          Second Owner  21.14 kmpl  1498 CC   
2     Individual       Manual           Third Owner   17.7 kmpl  1497 CC   
3     Individual       Manual           First Owner   23.0 kmpl  1396 CC   
4     Individual       Manual           First Owner   16.1 kmpl  1298 CC   
...          ...          ...                   ...         ...      ...   
7901  Individual       Manual           First Owner   18.5 kmpl  1197 CC   
7902  Individual       Manual  Fourth & Above Owner   16.8 kmpl  1493 CC   
7903  Individual       Manual           First Owner   19.3 kmpl  1248 CC   
7904  Individual       Manual           First Owner  23.57 kmpl  1396 CC   
7905  Individual       Manual           First Owner  23.57 kmpl  1396 CC   

       max_power                     torque  seats  
0         74 bhp             190Nm@ 2000rpm    5.0  
1     103.52 bhp        250Nm@ 1500-2500rpm    5.0  
2         78 bhp      12.7@ 2,700(kgm@ rpm)    5.0  
3         90 bhp   22.4 kgm at 1750-2750rpm    5.0  
4       88.2 bhp      11.5@ 4,500(kgm@ rpm)    5.0  
...          ...                        ...    ...  
7901   82.85 bhp           113.7Nm@ 4000rpm    5.0  
7902     110 bhp  24@ 1,900-2,750(kgm@ rpm)    5.0  
7903    73.9 bhp             190Nm@ 2000rpm    5.0  
7904      70 bhp        140Nm@ 1800-3000rpm    5.0  
7905      70 bhp        140Nm@ 1800-3000rpm    5.0  

[7819 rows x 13 columns]

In [298]:
#df['mileage_text'] = df['mileage'].str.extract('([A-Za-z]+)', expand=False)
#df['mileage_number'] = df['mileage'].str.extract('(\d+)', expand=False)
df_milage = df['mileage'].str.extract('([A-Za-z]+)', expand=False)
df_milage.value_counts()

kmpl    7819
km        87
Name: mileage, dtype: int64

In [299]:
## Definir si dejamos la misma unidad de medida aprox kmpl obviando la densidad (Aplicando una formula de conversion) -- FORMULA GOOGLE km/kg * 1.40
## De igual manera debemos definir si vamos a agragar columna de tipo de combustible

In [300]:
df[['mileage','mileage_unit']] = df['mileage'].str.split(' ', expand=True) ## Aun no aplicamos FORMULA solo sacamos texto

In [301]:
df.head()

name  year  selling_price  km_driven    fuel  \
0        Maruti Swift Dzire VDI  2014         450000     145500  Diesel   
1  Skoda Rapid 1.5 TDI Ambition  2014         370000     120000  Diesel   
2      Honda City 2017-2020 EXi  2006         158000     140000  Petrol   
3     Hyundai i20 Sportz Diesel  2010         225000     127000  Diesel   
4        Maruti Swift VXI BSIII  2007         130000     120000  Petrol   

  seller_type transmission         owner mileage   engine   max_power  \
0  Individual       Manual   First Owner    23.4  1248 CC      74 bhp   
1  Individual       Manual  Second Owner   21.14  1498 CC  103.52 bhp   
2  Individual       Manual   Third Owner    17.7  1497 CC      78 bhp   
3  Individual       Manual   First Owner    23.0  1396 CC      90 bhp   
4  Individual       Manual   First Owner    16.1  1298 CC    88.2 bhp   

                     torque  seats mileage_unit  
0            190Nm@ 2000rpm    5.0         kmpl  
1       250Nm@ 1500-2500rpm    5.0         kmpl  
2     12.7@ 2,700(kgm@ rpm)    5.0         kmpl  
3  22.4 kgm at 1750-2750rpm    5.0         kmpl  
4     11.5@ 4,500(kgm@ rpm)    5.0         kmpl

In [302]:
df[['engine','engine_unit']] = df['engine'].str.split(' ', expand=True)

In [303]:
unique_values, counts = np.unique(df['engine_unit'].values, return_counts=True)
unique_df = pd.DataFrame({'Value': unique_values, 'Count': counts})
print(unique_df)

  Value  Count
0    CC   7906


In [304]:
# Eliminar la columna original "engine_unit" pues todos son CC
df.drop("engine_unit", axis = 1, inplace=True)

In [305]:
df.head()

name  year  selling_price  km_driven    fuel  \
0        Maruti Swift Dzire VDI  2014         450000     145500  Diesel   
1  Skoda Rapid 1.5 TDI Ambition  2014         370000     120000  Diesel   
2      Honda City 2017-2020 EXi  2006         158000     140000  Petrol   
3     Hyundai i20 Sportz Diesel  2010         225000     127000  Diesel   
4        Maruti Swift VXI BSIII  2007         130000     120000  Petrol   

  seller_type transmission         owner mileage engine   max_power  \
0  Individual       Manual   First Owner    23.4   1248      74 bhp   
1  Individual       Manual  Second Owner   21.14   1498  103.52 bhp   
2  Individual       Manual   Third Owner    17.7   1497      78 bhp   
3  Individual       Manual   First Owner    23.0   1396      90 bhp   
4  Individual       Manual   First Owner    16.1   1298    88.2 bhp   

                     torque  seats mileage_unit  
0            190Nm@ 2000rpm    5.0         kmpl  
1       250Nm@ 1500-2500rpm    5.0         kmpl  
2     12.7@ 2,700(kgm@ rpm)    5.0         kmpl  
3  22.4 kgm at 1750-2750rpm    5.0         kmpl  
4     11.5@ 4,500(kgm@ rpm)    5.0         kmpl

In [306]:
df[['max_power','max_power_unit']] = df['max_power'].str.split(' ', expand=True)

In [307]:
unique_values, counts = np.unique(df['max_power_unit'].values, return_counts=True)
unique_df = pd.DataFrame({'Value': unique_values, 'Count': counts})
print(unique_df)

  Value  Count
0   bhp   7906


In [308]:
# Eliminar la columna original "max_power_unit" pues todos son bhp   
df.drop("max_power_unit", axis = 1, inplace=True)

In [309]:
df.head()

name  year  selling_price  km_driven    fuel  \
0        Maruti Swift Dzire VDI  2014         450000     145500  Diesel   
1  Skoda Rapid 1.5 TDI Ambition  2014         370000     120000  Diesel   
2      Honda City 2017-2020 EXi  2006         158000     140000  Petrol   
3     Hyundai i20 Sportz Diesel  2010         225000     127000  Diesel   
4        Maruti Swift VXI BSIII  2007         130000     120000  Petrol   

  seller_type transmission         owner mileage engine max_power  \
0  Individual       Manual   First Owner    23.4   1248        74   
1  Individual       Manual  Second Owner   21.14   1498    103.52   
2  Individual       Manual   Third Owner    17.7   1497        78   
3  Individual       Manual   First Owner    23.0   1396        90   
4  Individual       Manual   First Owner    16.1   1298      88.2   

                     torque  seats mileage_unit  
0            190Nm@ 2000rpm    5.0         kmpl  
1       250Nm@ 1500-2500rpm    5.0         kmpl  
2     12.7@ 2,700(kgm@ rpm)    5.0         kmpl  
3  22.4 kgm at 1750-2750rpm    5.0         kmpl  
4     11.5@ 4,500(kgm@ rpm)    5.0         kmpl